File: qc-application/src/main/java/com/eager/questioncloud/application/api/payment/QuestionPaymentController.java
package com.eager.questioncloud.application.api.payment;

import com.eager.questioncloud.application.api.common.DefaultResponse;
import com.eager.questioncloud.application.api.common.PagingResponse;
import com.eager.questioncloud.application.api.payment.QuestionPaymentControllerRequest.QuestionPaymentRequest;
import com.eager.questioncloud.application.security.UserPrincipal;
import com.eager.questioncloud.core.common.PagingInformation;
import com.eager.questioncloud.core.domain.payment.dto.QuestionPaymentHistory;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/payment/question")
@RequiredArgsConstructor
public class QuestionPaymentController {
    private final QuestionPaymentService questionPaymentService;
    private final QuestionPaymentHistoryService questionPaymentHistoryService;

    @PostMapping
    @ApiResponses(value = {
        @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "요청 성공")
    })
    @Operation(operationId = "문제 구매", summary = "문제 구매", tags = {"question-payment"}, description = "문제 구매")
    public DefaultResponse questionPayment(@AuthenticationPrincipal UserPrincipal userPrincipal, @RequestBody QuestionPaymentRequest request) {
        questionPaymentService.payment(userPrincipal.getUser().getUid(), request.getQuestionIds(), request.getUserCouponId());
        return DefaultResponse.success();
    }

    @GetMapping("/history")
    @ApiResponses(value = {
        @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "요청 성공")
    })
    @Operation(operationId = "문제 구매 내역 조회", summary = "문제 구매 내역 조회", tags = {"question-payment"}, description = "문제 구매 내역 조회")
    public PagingResponse<QuestionPaymentHistory> getQuestionPaymentHistory(
        @AuthenticationPrincipal UserPrincipal userPrincipal, PagingInformation pagingInformation) {
        return new PagingResponse<>(
            questionPaymentHistoryService.countQuestionPaymentHistory(userPrincipal.getUser().getUid()),
            questionPaymentHistoryService.getQuestionPaymentHistory(userPrincipal.getUser().getUid(), pagingInformation));
    }
}

File: qc-core/qc-pay/qc-pay-api/src/main/kotlin/com/eager/questioncloud/payment/question/controller/QuestionPaymentController.kt
package com.eager.questioncloud.payment.question.controller

import com.eager.questioncloud.common.auth.UserPrincipal
import com.eager.questioncloud.common.dto.DefaultResponse
import com.eager.questioncloud.common.dto.PagingResponse
import com.eager.questioncloud.common.pagination.PagingInformation
import com.eager.questioncloud.payment.domain.QuestionPaymentHistory
import com.eager.questioncloud.payment.question.command.QuestionPaymentCommand
import com.eager.questioncloud.payment.question.dto.QuestionPaymentRequest
import com.eager.questioncloud.payment.question.service.QuestionPaymentHistoryService
import com.eager.questioncloud.payment.question.service.QuestionPaymentService
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/payment/question")
class QuestionPaymentController(
    private val questionPaymentService: QuestionPaymentService,
    private val questionPaymentHistoryService: QuestionPaymentHistoryService,
) {
    @PostMapping
    fun questionPayment(
        userPrincipal: UserPrincipal,
        @RequestBody request: QuestionPaymentRequest
    ): DefaultResponse {
        questionPaymentService.payment(
            QuestionPaymentCommand(
                userPrincipal.userId,
                request.orders.map { it.toCommand() },
                request.userCouponId
            )
        )
        return DefaultResponse.success()
    }
    
    @GetMapping("/history")
    fun getQuestionPaymentHistory(
        userPrincipal: UserPrincipal, pagingInformation: PagingInformation
    ): PagingResponse<QuestionPaymentHistory> {
        return PagingResponse(
            questionPaymentHistoryService.countQuestionPaymentHistory(userPrincipal.userId),
            questionPaymentHistoryService.getQuestionPaymentHistory(userPrincipal.userId, pagingInformation)
        )
    }
}
